Convert SDK to TypeScript

nan-li opened the pull request:
❗ Note: If you look at [all files changed](https://github.com/OneSignal/react-native-onesignal/pull/1312/files), it **doesn't** accurately show the changes from:
`index.js` → `index.ts` 
Refer to [this commit](https://github.com/OneSignal/react-native-onesignal/pull/1312/commits/61dc879fa8c45fdf5963aaa23edbd8bf08e38134) for a more accurate representation.

# Description
## One Line Summary
Migrate the SDK JavaScript files under `src/` to TypeScript. 

## Details
### Motivation
TypeScript can support writing code that is less likely to contain bugs and is easier to maintain.

### Scope
No new methods are added or removed but types may potentially cause changes for users using different types than those specified now. These changes should also be thoroughly tested.

# Testing
## Unit testing
N/A

## Manual testing
- [x] Ran example app on Android and iOS emulator
- [x] Tested access of all public models such as `OSNotification`, `OutcomeEvent`, `LogLevel`, etc.
- [x] Tested access and setting of all the handlers and observers, and the event type received by the handler 

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [X] Public API changes - in a way

# Checklist
## Overview
   - [X] I have filled out all **REQUIRED** sections above
   - [X] PR does one thing
   - [X] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [X] I have included test coverage for these changes, or explained why they are not needed
   - [X] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [X] Code is as readable as possible.
   - [X] I have reviewed this PR myself, ensuring it meets each checklist item

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/onesignal/react-native-onesignal/1312)
<!-- Reviewable:end -->

nan-li committed:
Create `tsconfig.json`

nan-li committed:
Update `.gitignore` to exclude `dist` directory

nan-li committed:
Rename `events.js` to `events.ts`

- ran `tsc` afterwords and nothing to be done

nan-li committed:
Rename `OSNotification.js` to `OSNotification.ts`

- add OSNotification fields (from index.d.ts)
- make receivedEvent of type OSNotification based on shape
- fix typo in `lockScreenVisibility`

nan-li committed:
Rename `NotificationReceivedEvent.js` to `.ts`

- add OSNotification field and some typing

nan-li committed:
Moves types from `index.d.ts` to own files

Make dir /models
- Subscription.ts
- InAppMessage.ts
- NotificationEvents.ts
- Outcomes.ts

nan-li committed:
Rename `helpers.js` to `.ts`

- add typings for isValidCallback, isObjectNonNull, and isMultipleInstancesPossible

nan-li committed:
Rename `index.js` to `.ts` (Step 1/2)

- In this commit, I have renamed `index.js` to `index.ts` but have made no further changes than simply renaming the file extension
- In the next commit, I will include the changes to the TypeScript file
- This is done in 2 steps because I was unable to rename and change a file in one commit. My commit kept coming out as 1 file deleted, 1 file added.

nan-li committed:
Add typings to `index.ts` (Step 2/2)

- This is the 2nd commit to follow the previous commit
- In this commit, I add the types to `index.ts`
- I primarily followed the existing type annotations in `index.d.ts` with some changes

nan-li committed:
Rename `EventManager.js` to `.ts`

- Rename and add typings

brettshwom-onesignal commented on src/helpers.ts:
```diff
@@ -6,22 +6,22 @@ import {
 }
 
-export function isObjectNonNull(object) {
+export function isObjectNonNull(object: any) {
```
how about using a type guard?

https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

```
function isObjectNonNull<T extends any>(object: T): obj is Exclude<T, null> {
    return object !== null;
}
```

nan-li replied:
Also excluded `undefined`?

brettshwom-onesignal commented on src/helpers.ts:
```diff
@@ -6,22 +6,22 @@ import {
 } from './events';
 
-export function isValidCallback(handler) {
+export function isValidCallback(handler: any) { // type will be checked in this method
```
invariant throws an Error when the predicate fails right?

if so, how about using an assertion function: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions
```
function isValidCallback(x: any): asserts x is () => void {
    if (typeof x !== 'function') {
        throw 'not a valid fucntion'
    }
}
```

with invariant instead:
```
function isValidCallback(x: any): asserts x is () => void {
     invariant(
        typeof handler === 'function',
        'Must provide a valid callback'
    );
}
````

rgomezp replied:
can be more specific. a handler will always be of type `Function`

nan-li replied:
Done, made handler of type `Function`

brettshwom-onesignal commented:
nice!

rgomezp commented on tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "compilerOptions": {
+    "outDir": "./built",
+    "target": "es6",
```
target should be es5

nan-li replied:
Done

nan-li replied:
The current `jsconfig` has target as es5. Is this different because of javascript v. typescript?

rgomezp commented on tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+    "examples",
+    "ios"
+  ]
+}
```
newline

nan-li replied:
Done

rgomezp commented on tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+    "moduleResolution": "node",
+    "noImplicitAny": true,
+    "lib": [
+      "es6", 
```
this is fine...
https://stephenweiss.dev/typescript-compiler-lib-vs-target/

rgomezp commented on .gitignore:
```diff
@@ -29,3 +29,4 @@ ios/RCTOneSignal.xcodeproj/project.xcworkspace/xcshareddata
 examples/RNOneSignal/ios/Pods/Pods.xcodeproj/xcuserdata
 android/.idea
 images
+built/
```
newline

nan-li replied:
Done, and changed `built/` to `dist/`

rgomezp commented on src/NotificationReceivedEvent.ts:
```diff
@@ -3,11 +3,13 @@ import OSNotification from './OSNotification';
     }
 
-    complete(notification) {
+    complete(notification: OSNotification) {
```
Should be optional

rgomezp replied:
also missing return type

nan-li replied:
Sorry, I don't know why I split up this file into 2 commits. This was updated in a later commit (Update `NotificationReceivedEvent` typings).

rgomezp commented on src/NotificationReceivedEvent.ts:
```diff
@@ -3,11 +3,13 @@ import OSNotification from './OSNotification';
 const RNOneSignal = NativeModules.OneSignal;
 
 export default class NotificationReceivedEvent {
```
`getNotification` function missing return type

nan-li replied:
Sorry, I don't know why I split up this file into 2 commits. This was updated in a later commit (Update NotificationReceivedEvent typings).

rgomezp commented on types/InAppMessage.ts:
```diff
@@ -0,0 +1,19 @@
+export interface InAppMessage {
```
Prefer to use `models` as directory name

nan-li replied:
Done

rgomezp commented on src/OSNotification.ts:
```diff
@@ -1,7 +1,43 @@
+        mutableContent      ?: boolean;
+        contentAvailable    ?: string;
+        relevanceScore      ?: number;
+        interruptionLevel   ?: string;
```
indentation is off

nan-li replied:
I used the indentation from existing `index.d.ts` file, I think to show the `ios` and `android` sections better. I can line up the indentation if that is preferable.

rgomezp commented on src/NotificationReceivedEvent.ts:
```diff
@@ -23,7 +23,7 @@ export default class NotificationReceivedEvent {
+    getNotification() : OSNotification {
         return this.notification;
     }
 }
```
newline

nan-li replied:
Done

rgomezp commented on src/index.ts:
```diff
@@ -15,72 +15,125 @@ import {
 const eventManager = new EventManager(RNOneSignal);
 
+// 0 = None, 1 = Fatal, 2 = Errors, 3 = Warnings, 4 = Info, 5 = Debug, 6 = Verbose
+export type LogLevel = 0 | 1 | 2 | 3 | 4 | 5 | 6;
```
unless used elsewhere, no need to export

nan-li replied:
Makes sense.. done

rgomezp commented on src/index.ts:
```diff
@@ -93,16 +146,21 @@ export default class OneSignal {
      *  - notification data
      *  - `complete` function that accepts the `NotificationReceivedEvent`
-     * @param  {function} handler
+     * @param  {(event:NotificationReceivedEvent)=>void} handler
```
spaces around =>

nan-li replied:
These were copied over from existing `index.d.ts` file and there's a bunch of them without spaces. Should I modify all?

nan-li replied:
Done, put spaces around all the functions needing it

rgomezp commented on src/index.ts:
```diff
@@ -123,7 +186,16 @@ export default class OneSignal {
+     * @param  {(response:boolean)=>void} handler
+     * @returns void
+     */
+    static registerForProvisionalAuthorization(handler?: (response: boolean) => void): void {
```
callback should be required since `isValidCallback` would throw anyway

nan-li replied:
Done

rgomezp commented on src/index.ts:
```diff
@@ -152,19 +234,32 @@ export default class OneSignal {
+     * True if the application has location share activated, false otherwise
+     * @returns Promise<boolean>
+     */
+    static isLocationShared(): Promise<boolean | void> {
```
this should not return void. should either be true or false. if issue in the bridge, this should also be fixed

nan-li replied:
In the body, if the native module `RNOneSignal` is `null`, this method doesn't call native code and returns `Promise.resolve()` with no arguments, and because of that, I added `void` to the return type: 

> Type 'Promise\<void>' is not assignable to type 'Promise\<boolean>'.

That is incorrect? Then, (1) should I resolve the no-parameter `Promise` with a value instead? Or (2) should this be a `reject` instead?

If (1), I can resolve with `false` for booleans, but for the methods `getDeviceState` and `getTriggerValueForKey`, it may be misleading to resolve to `null`. In the bridge, the promise is rejected if `getDeviceState` returns nothing but for `getTriggerValueForKey`, only in the iOS bridge is the promise rejected if value returned is null.

nan-li replied:
Done, resolve to `false`

rgomezp commented on src/index.ts:
```diff
@@ -175,8 +270,10 @@ export default class OneSignal {
+     * @returns Promise<DeviceState>
      */
-    static async getDeviceState() {
+    static async getDeviceState(): Promise<DeviceState | void> {
```
this should not return void. should return Device State or `null`. make sure comment return type matches.

nan-li replied:
Done

rgomezp commented on src/index.ts:
```diff
@@ -188,15 +285,26 @@ export default class OneSignal {
+     * @param  {string} value
+     * @returns void
+     */
+    static sendTag(key: string, value: string | boolean): void {
```
missing `number` type from value

rgomezp replied:
test this to ensure no issues in bridge

nan-li replied:
In the [iOS SDK](https://github.com/OneSignal/OneSignal-iOS-SDK/blob/fd06483b314f842c16e0a6f74dfb27a04dc7df66/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L1353) and [Android SDK](https://github.com/OneSignal/OneSignal-Android-SDK/blob/54e4b0fe45bdb7168788d2072a98d573e5a3e599/OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java#L1813) and OneSignal documentation (https://documentation.onesignal.com/docs/data-tag-implementation#sendtag-method), value is `string`. Another documentation page on tags suggests numbers can be used and the example passes the number as a string version.

I added `boolean` to the value type here because in the body of this method, we are letting users pass in a `boolean`.
```
        if (typeof value === "boolean") {
            value = value.toString();
        }
```
❓ Should I add `number` here as well and convert it `toString`. OR, remove the `boolean` type and only accept string values?

In Flutter, value is a `dynamic` type, and in Cordova, there is no type (except the implementation file says `string`).

nan-li replied:
Done.. `tags` should only have `string` values

rgomezp commented on src/index.ts:
```diff
@@ -210,7 +318,12 @@ export default class OneSignal {
+     * @param  {object} tags
+     * @returns void
+     */
+    static sendTags(tags: { [key: string]: any }): void {
```
any should be same type union as `sendTag`

nan-li replied:
Done, value is `string`

rgomezp commented on src/index.ts:
```diff
@@ -400,22 +595,30 @@ export default class OneSignal {
+     * @param  {string} value
+     * @returns void
+     */
+    static addTrigger(key: string, value: string): void {
```
value can be number or bool as well https://documentation.onesignal.com/docs/iam-sdk-methods#addtrigger-method

nan-li replied:
That's right.. Done.

rgomezp commented on src/index.ts:
```diff
@@ -400,22 +595,30 @@ export default class OneSignal {
         }
 
-        let trigger = {};
+        let trigger: { [key: string]: string } = {};
```
value can be number or bool as well https://documentation.onesignal.com/docs/iam-sdk-methods#addtrigger-method

nan-li replied:
Done

rgomezp commented on src/index.ts:
```diff
@@ -400,22 +595,30 @@ export default class OneSignal {
+     * @param  {object} triggers
+     * @returns void
+     */
+    static addTriggers(triggers: object): void {
```
type definition can be made more specific (see comment above re: `addTrigger`)

nan-li replied:
Done.. type is `{ [key: string]: string | number | boolean }`

rgomezp commented on src/index.ts:
```diff
@@ -427,82 +630,150 @@ export default class OneSignal {
+    /**
+     * Gets a trigger value for a provided trigger key.
+     * @param  {string} key
+     * @returns Promise<string>
```
make sure returns on all comments match the actual return values

nan-li replied:
Done

rgomezp commented on src/index.ts:
```diff
@@ -427,82 +630,150 @@ export default class OneSignal {
+     * @param  {(event:OutcomeEvent)=>void} handler
+     * @returns void
+     */
+    static sendOutcom(name: string, handler?: (event: OutcomeEvent) => void): void {
```
typo

rgomezp replied:
double check all functions

nan-li replied:
😔 Fixed typo. Yes, will followup with testing all functions.

rgomezp commented on src/index.ts:
```diff
@@ -427,82 +630,150 @@ export default class OneSignal {
+     * Did the user provide privacy consent for GDPR purposes.
+     * @returns Promise<boolean>
+     */
+    static userProvidedPrivacyConsent(): Promise<boolean | void> {
```
Should not return void. True or False (void ~> default to false)

nan-li replied:
Done, default resolve to `false`

rgomezp commented on src/index.ts:
```diff
@@ -427,82 +630,150 @@ export default class OneSignal {
+     * True if the application requires user privacy consent, false otherwise
+     * @returns Promise<boolean>
+     */
+    static requiresUserPrivacyConsent(): Promise<boolean | void> {
```
Should not return void. True or False (void ~> default to false)

nan-li replied:
Done, resolve to `false`

rgomezp commented on src/EventManager.ts:
```diff
@@ -82,13 +89,13 @@ export default class EventManager {
-    generateEventListener(eventName) {
-        const addListenerCallback = (payload) => {
+    generateEventListener(eventName: string) {
+        const addListenerCallback = (payload: any) => {
```
can we make `payload` type more specific?

you may need to create a new variable `finalPayload` below to avoid type errors

nan-li replied:
`payload` is one of the 11 event types:

> 4 ChangeEvents like `PermissionChange`
3 other events (`NotificationReceivedEvent` `OpenedEvent` `InAppMessageAction`) + 4 IAM lifecycle events

First, I should figure out how to use the event types you brought up above in the methods `addEventHandler` and `setEventHandler`.

nan-li replied:
Done. `payload` is `Object`

rgomezp commented on src/EventManager.ts:
```diff
@@ -112,7 +119,7 @@ export default class EventManager {
     }
 
-    getFinalPayload(eventName, payload) {
+    getFinalPayload(eventName: string, payload: any) {
```
missing return type

nan-li replied:
Done. `payload` is `Object` and so is return type

rgomezp commented on src/EventManager.ts:
```diff
@@ -82,13 +89,13 @@ export default class EventManager {
     // returns an event listener with the js to native mapping
-    generateEventListener(eventName) {
-        const addListenerCallback = (payload) => {
+    generateEventListener(eventName: string) {
```
missing return type

nan-li replied:
Done

rgomezp commented on src/EventManager.ts:
```diff
@@ -82,13 +89,13 @@ export default class EventManager {
      * @param  {function} handler
      */
-    clearEventHandler(eventName) {
+    clearEventHandler(eventName: string) {
```
missing return type

nan-li replied:
Done

rgomezp commented on src/EventManager.ts:
```diff
@@ -72,7 +79,7 @@ export default class EventManager {
      * @param  {function} handler
      */
-    addEventHandler(eventName, handler) {
+    addEventHandler(eventName: string, handler: (event: any) => void): void {
```
`event` type should be `ChangeEvent<ChangeObject>` where 

`export type ChangeObject = PermissionChange | SubscriptionChange | EmailSubscriptionChange | SMSSubscriptionChange;`

nan-li replied:
I tried to use more specific event types for these `ChangeEvent` as well as the other class of events 

> `export type HandlerEvent = NotificationReceivedEvent | OpenedEvent | InAppMessageAction` 

I mentioned a little bit in the Paper that I didn't get it working.
This is error I see in `index.ts`

![Screen Shot 2021-11-29 at 12 56 08 PM](https://user-images.githubusercontent.com/4022291/143942804-ed0b3c2c-f659-40c3-a77b-e244285f3c9d.png)

nan-li replied:
Done, `event` is `ChangeEvent<T>`

rgomezp commented on tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+
+{
+  "compilerOptions": {
+    "outDir": "./built",
```
outDir should be `dist` or `build`

nan-li replied:
Done, changed to `dist`. 
I had followed [TS Handbook Migrating from JavaScript
](https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html) for the name previously.

rgomezp reviewed (commented):
Several additional notes:
* the type definitions should be in a `models` subdirectory in `src`
* put all events related files in a subdirectory `events`

Delete `index.d.ts`
* This should be auto-generated by the typescript compiler and placed in the `dist` directory
* Update `package.json` to use the `typings` location of the generated file

Add build and prepare scripts to `package.json`
* The build script should run the typescript compiler and put the out files in the `outDir` specified in the tsconfig file
* The prepare script will be autorun by the npm publish command
* The build script should be called by the prepare script

Add `dist` to .gitignore (but not .npmignore)

nan-li committed:
Put events-related files in a subdirectory `events/`

- EventManager.ts
- events.ts
- NotificationReceivedEvent.ts

nan-li committed:
Remove `jsconfig.json`

nan-li committed:
Update package.json

- main and types use files built in `dist`
- prepare and build scripts
- add types in devDependencies

nan-li commented:
Some questions or notes I still have:
1. How best to make the types under `models` directory available to users? For example, that they can import and have the type of OutcomeEvent such as this in the example app: 

> `import { OutcomeEvent } from 'react-native-onesignal'`

2. TODO: the example app will need to be updated. For example, the example app imports `OutcomeEvent` just like above but now doesn't have access unless getting from 'react-native-onesignal/dist/models/Outcomes'. Also some methods like `getDeviceState` can now return `null`, so need to check before accessing its properties.

3. Implementation files are generated for **all** files and not only `index.ts` because I added  `"declaration": true` to tsconfig. I think this is somewhat correct because those types are needed by index.

4. I think it may be preferable to send a rejected Promise when RNOneSignal is `null` when users call those methods, such as `getDeviceState` or `getTriggerValueForKey` since it may be misleading to receive `null` and a user will think the key doesn't exist when it may exist.

5. I yarn added my local repository to a typescript-based demo app and seems to be working: method signatures link to `dist/index.d.ts` in node_modules and sending a tag shows in dashboard.

rgomezp commented:
> Some questions or notes I still have:
> 
> 1. How best to make the types under `models` directory available to users? For example, that they can import and have the type of OutcomeEvent such as this in the example app:
> 
> > `import { OutcomeEvent } from 'react-native-onesignal'`
> 

These should be generated automatically and added to the typings file if you configured your tsconfig right. `isolatedModules` should be false. Just make sure you export all of the models in index.ts. Ex:

```
export {
  InAppMessage,
  InAppMessageAction,
  InAppMessageLifecycleHandlerObject,
  OpenedEvent,
  OutcomeEvent,
  NotificationReceivedEvent,
  DeviceState,
  ChangeEvent,
  PermissionChange,
  SubscriptionChange,
  EmailSubscriptionChange,
  SMSSubscriptionChange,
  OSNotification
  ...
  <other public models>
}
```

> 2. TODO: the example app will need to be updated. For example, the example app imports `OutcomeEvent` just like above but now doesn't have access unless getting from 'react-native-onesignal/dist/models/Outcomes'. Also some methods like `getDeviceState` can now return `null`, so need to check before accessing its properties.

Feel free to make changes to the Example app with these changes.

> 3. Implementation files are generated for **all** files and not only `index.ts` because I added  `"declaration": true` to tsconfig. I think this is somewhat correct because those types are needed by index.

See above point in 1). We need to make sure all public models get exported so they can be used in a project.

> 4. I think it may be preferable to send a rejected Promise when RNOneSignal is `null` when users call those methods, such as `getDeviceState` or `getTriggerValueForKey` since it may be misleading to receive `null` and a user will think the key doesn't exist when it may exist.

We check if `RNOneSignal` is null on all our functions even if it is a synchronous function not returning a promise. Perhaps we could actually throw with a good error message. Let's encapsulate this in a helper function so as to not clutter the index functions. Something like `checkNativeModuleLoaded(RNOneSignal)` or something similar.

> 5. I yarn added my local repository to a typescript-based demo app and seems to be working: method signatures link to `dist/index.d.ts` in node_modules and sending a tag shows in dashboard.

Double check imported types are working (e.g: OSNotification in the `setNotificationWillShowInForegroundHandler` complete function) and you can import them from your project like:

```
const myNotification: OSNotification = { ... }
```

Make sure we're not forgetting to export any important models as named exports.

rgomezp commented on src/index.d.ts:
```diff
@@ -20,26 +20,26 @@ declare module 'react-native-onesignal' {
         hasPrompted             ?: boolean;   // ios
```
This file should not even be here since it should be generated automatically. I think this commit is unnecessary

nan-li replied:
Done, removed this commit

rgomezp commented on src/NotificationReceivedEvent.ts:
```diff
@@ -3,13 +3,13 @@ import OSNotification from './OSNotification';
 
 export default class NotificationReceivedEvent {
-    notification            : OSNotification; //TODO: Can this be optional?
+    notification            : OSNotification;
```
unnecessary spacing

nan-li replied:
Done

rgomezp commented on src/NotificationReceivedEvent.ts:
```diff
@@ -3,13 +3,13 @@ import OSNotification from './OSNotification';
     }
 
-    complete(notification: OSNotification) {
+    complete(notification?: OSNotification) : void {
```
remove space after )

nan-li replied:
Done

rgomezp commented on src/NotificationReceivedEvent.ts:
```diff
@@ -23,7 +23,7 @@ export default class NotificationReceivedEvent {
     }
 
-    getNotification() {
+    getNotification() : OSNotification {
```
Remove space after )

nan-li replied:
Done

rgomezp commented on src/NotificationReceivedEvent.ts:
```diff
@@ -3,13 +3,13 @@ import OSNotification from './OSNotification';
 const RNOneSignal = NativeModules.OneSignal;
```
imported `Platform` is unused

rgomezp replied:
add `"noUnusedLocals": true` to tsconfig to catch these

nan-li replied:
Done

rgomezp commented on src/index.ts:
```diff
@@ -0,0 +1,771 @@
+const eventManager = new EventManager(RNOneSignal);
+
+// 0 = None, 1 = Fatal, 2 = Errors, 3 = Warnings, 4 = Info, 5 = Debug, 6 = Verbose
+type LogLevel = 0 | 1 | 2 | 3 | 4 | 5 | 6;
```
we should export this value

nan-li replied:
Done

rgomezp commented on tsconfig.json:
```diff
@@ -0,0 +1,22 @@
+      "es6"
+    ],
+    "declaration": true,
+    "isolatedModules": true,
```
change to false (https://www.typescriptlang.org/tsconfig#isolatedModules)

nan-li replied:
Done, removed since default is false

rgomezp commented:
Nit: there are trailing spaces all over the project. Install this plugin to help you identify them. https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces

rgomezp commented on tsconfig.json:
```diff
@@ -0,0 +1,22 @@
+    "node_modules",
+    "android",
+    "examples",
+    "ios"
```
add "dist", then you don't have to remove the dist directory every time you rebuild (mentioned in other comment)

nan-li replied:
Done

nan-li committed:
Delete `index.d.ts`

nan-li committed:
Export all public models in `index.ts`

nan-li committed:
Update Demo app

- After making the TS changes, `getDeviceState` can return `null`, so we need to check this before accessing its properties

- addSMSSubscriptionObserver to the demo app as well since this is the only one missing

nan-li commented:
1. > We check if `RNOneSignal` is null on all our functions even if it is a synchronous function not returning a promise. Perhaps we could actually throw with a good error message. Let's encapsulate this in a helper function so as to not clutter the index functions. Something like `checkNativeModuleLoaded(RNOneSignal)` or something similar.

I was testing this out and if this method throws an error, users will have to go back and surround all their OneSignal calls with a try-catch or equivalent. Is that something we want to do? Or leave it as it is.. or something else.

2. All the public models are exported in `index.ts` and I checked that each one is accessible from demo application by importing from 'react-native-onesignal' and the properties are correct on the types. Also checked the handlers and observers and their parameter type and properties.

3. Changed some `getDeviceState` calls in the demo app because the method can now return `null`, so we need to check this before accessing properties of `DeviceState` such as `userId`. This may similarly affect existing users of the SDK.

rgomezp commented on src/index.ts:
```diff
@@ -0,0 +1,792 @@
+
+    /**
+     * Completes OneSignal initialization by setting the OneSignal Application ID.
+     * 1/2 of the initialization process.
```
Remove this line

nan-li replied:
done

rgomezp commented on src/events/NotificationReceivedEvent.ts:
```diff
@@ -1,13 +1,15 @@
 
 export default class NotificationReceivedEvent {
-    constructor(receivedEvent){
+    notification: OSNotification;
```
should be private

nan-li replied:
done

rgomezp commented on src/events/EventManager.ts:
```diff
@@ -30,17 +32,24 @@ const eventList = [
+    oneSignalEventEmitter: NativeEventEmitter;
+    eventHandlerMap: Map<string, (event: any) => void>;
+    eventHandlerArrayMap: Map<string, Array<(event: any) => void>>;
+    listeners: { [key: string]: EmitterSubscription };
```
all these properties should be private

nan-li replied:
done

rgomezp commented:
> I was testing this out and if this method throws an error, users will have to go back and surround all their OneSignal calls with a try-catch or equivalent. Is that something we want to do? Or leave it as it is.. or something else.

That's a good consideration. Alternatively, we could do a `console.error`.

```ts
function isNativeModuleLoaded(RNOneSignal) {
   if (!isObjectNonNull(RNOneSignal) {
      console.error("Could not load RNOneSignal native module. Make sure native dependencies are properly linked.");
      return false;
   }
   return true;
}
```

In index:

```ts
if (!isNativeModuleLoaded(RNOneSignal)) return;
```

Or something similar.

nan-li committed:
Use helper function to log error message if Native Module is null

rgomezp reviewed (approved):
LGTM

brettshwom-onesignal commented on examples/RNOneSignalTS/src/OSDemo.tsx:
```diff
@@ -90,8 +93,9 @@ class OSDemo extends React.Component<Props, State> {
-            name : state.emailAddress,
-            isSubscribed : state.isSubscribed
+            // state is possibly 'null' so need to check
+            name : state? state.emailAddress : "",
```
very minor (i.e. this suggestion shouldnt block the merging of this PR)

how about using nullish coalescence and optional chaining?


https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#nullish-coalescing

```
state?.emailAddress ?? ""
```

nan-li replied:
Good points, thanks. I'll make a followup PR for the example app.
